src/device.rs

```diff
@@ -42,3 +49,127 @@ impl Drop for Device {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
+        match self {
+            Isa::Sse2 => write!(f, "sse2"),
+            Isa::Sse4_2 => write!(f, "sse4.2"),
+            Isa::Avx => write!(f, "avx"),
+            Isa::Avx2 => write!(f, "avx2"),
+            Isa::Avx512 => write!(f, "avx512"),
+        }
+    }
+}
+
+/// Frequency level of the application.
+#[derive(Debug, Clone, Copy)]
+pub enum FrequencyLevel {
+    /// Run at highest frequency.
+    Simd128,
+
+    /// Run at AVX2-heavy frequency level.
+    Simd256,
+
+    /// Run at AVX512-heavy frequency level.
+    Simd512,
+}
+
+impl Display for FrequencyLevel {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
+        match self {
+            FrequencyLevel::Simd128 => write!(f, "simd128"),
+            FrequencyLevel::Simd256 => write!(f, "simd256"),
+            FrequencyLevel::Simd512 => write!(f, "simd512"),
+        }
+    }
+}
+
+/// Embree device configuration.
+pub struct Config {
+    /// The number of build threads. A value of 0 enables all detected hardware
+    /// threads. By default all hardware threads are used.
+    pub threads: u32,
+
+    /// The number of user threads that can be used to join and participate in a
+    /// scene commit using `rtcJoinCommitScene`.
+    pub user_threads: u32,
+
+    /// Whether build threads are affinitized to hardware threads. This is disabled
+    /// by default on standard CPUs, and enabled by default on Xeon Phi Processors.
+    pub set_affinity: bool,
+
+    /// When enabled, the build threads are started upfront. Useful for benchmarking
+    /// to exclude thread creation time. This is disabled by default.
+    pub start_threads: bool,
+
+    /// ISA selection. By default the ISA is chosen automatically.
+    pub isa: Isa,
+
+    /// Configures the automated ISA selection to use maximally the specified ISA.
+    pub max_isa: Isa,
+
+    /// Enables or disables usage of huge pages. Enabled by default under Linux but
+    /// disabled by default on Windows and macOS.
+    pub hugepages: bool,
+
+    /// Enables or disables the SeLockMemoryPrivilege privilege which is required to
+    /// use huge pages on Windows. This option has only effect on Windows and is ignored
+    /// on other platforms.
+    pub enable_selockmemoryprivilege: bool,
+
+    /// Verbosity of the output [0, 1, 2, 3]. No output when set to 0. The higher the
+    /// level, the more the output. By default the output is set to 0.
+    pub verbose: u32,
+
+    /// Frequency level the application want to run on. See [`FrequencyLevel`].
+    /// When some frequency level is specified, Embree will avoid doing optimizations
+    /// that may reduce the frequency level below the level specified.
+    pub frequency_level: FrequencyLevel,
+}
+
+impl Config {
+    /// Converts the configuration to a C string.
+    pub fn to_c_string(&self) -> CString {
+        let formated = format!("threads={},verbose={},set_affinity={},start_threads={},\
```

Comment:
Here'd we'd only want to include the isa and frequency level options if they were set to actual values by the application, because ISA = SSE2 is not equivalent to specifying no ISA, and same for frequency level = simd256 not being the same as specifying nothing and letting Embree pick.